Gradle conversion issue: Unsupported class file major version 69

Hi

We have Gradle project that uses JDK 25 and Gradle 9.3.1
We tried to migrate it to Mill however _mill init_ command resulted in this error (I removed non-important parts of stack-trace):

```
converting Gradle build
Downloading https://services.gradle.org/distributions/gradle-8.11.1-bin.zip
..................................................................................................................................
org.gradle.tooling.BuildException: Could not fetch model of type 'BuildModel' using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.11.1-bin.zip'.
        at org.gradle.tooling.internal.consumer.ConnectionExceptionTransformer.transform(ConnectionExceptionTransformer.java:53)
        at org.gradle.tooling.internal.consumer.ResultHandlerAdapter.onFailure(ResultHandlerAdapter.java:42)
        at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor$1$1.run(DefaultAsyncConsumerActionExecutor.java:68)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.gradle.tooling.internal.consumer.BlockingResultHandler.getResult(BlockingResultHandler.java:46)
        at org.gradle.tooling.internal.consumer.DefaultModelBuilder.get(DefaultModelBuilder.java:50)
        at mill.main.gradle.MillGradleBuildGenMain$.$anonfun$9(MillGradleBuildGenMain.scala:69)
        at scala.util.Using$.resource(Using.scala:298)
        at mill.main.gradle.MillGradleBuildGenMain$.init(MillGradleBuildGenMain.scala:71)
        at mill.main.gradle.MillGradleBuildGenMain$.$anonfun$4(MillGradleBuildGenMain.scala:17)
        at mainargs.Invoker$.$anonfun$2(Invoker.scala:77)
        at mainargs.Result.flatMap(Result.scala:13)
        at mainargs.Result.flatMap$(Result.scala:7)
        at mainargs.Result$Success.flatMap(Result.scala:23)
        at mainargs.Invoker$.invoke0(Invoker.scala:77)
        at mainargs.Invoker$.invoke(Invoker.scala:86)
        at mainargs.Invoker$.invokeLocal$1(Invoker.scala:106)
        at mainargs.Invoker$.groupArgs$1$$anonfun$2(Invoker.scala:121)
        at mainargs.Result.flatMap(Result.scala:13)
        at mainargs.Result.flatMap$(Result.scala:7)
        at mainargs.Result$Success.flatMap(Result.scala:23)
        at mainargs.Invoker$.groupArgs$1(Invoker.scala:121)
        at mainargs.Invoker$.runMains(Invoker.scala:126)
        at mainargs.ParserForMethods.runRaw0(Parser.scala:291)
        at mainargs.ParserForMethods.runEither(Parser.scala:179)
        at mainargs.ParserForMethods.runOrExit(Parser.scala:73)
        at mill.main.gradle.MillGradleBuildGenMain$.main(MillGradleBuildGenMain.scala:17)
        at mill.main.gradle.MillGradleBuildGenMain.main(MillGradleBuildGenMain.scala)
Caused by: org.gradle.internal.exceptions.LocationAwareException: BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 69
        at org.gradle.initialization.exception.DefaultExceptionAnalyser.transform(DefaultExceptionAnalyser.java:97)
        at org.gradle.initialization.exception.DefaultExceptionAnalyser.collectFailures(DefaultExceptionAnalyser.java:63)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
Caused by: BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 69
        at org.gradle.groovy.scripts.internal.DefaultScriptCompilationHandler.compileScript(DefaultScriptCompilationHandler.java:160)
        at org.gradle.groovy.scripts.internal.DefaultScriptCompilationHandler.compileToDir(DefaultScriptCompilationHandler.java:112)
        at org.gradle.groovy.scripts.internal.BuildOperationBackedScriptCompilationHandler$2.run(BuildOperationBackedScriptCompilationHandler.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$fromBuildModel$2(DefaultBuildTreeLifecycleController.java:90)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$runBuild$4(DefaultBuildTreeLifecycleController.java:120)
        ... 75 more
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 69
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:200)
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:180)
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:166)
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:288)
```

We were surprised that Mill uses such old Gradle version (8.11.1) that released on October, 2024. It seems that using Gradle 9.x for migration could resolve this issue but the only article about automatic migration has no information about that: https://mill-build.org/mill/migrating/auto-migrating.html

--- a/libs/init/gradle/src/mill/main/gradle/MillGradleBuildGenMain.scala
+++ b/libs/init/gradle/src/mill/main/gradle/MillGradleBuildGenMain.scala
@@ -8,6 +8,7 @@ import org.gradle.tooling.GradleConnector
 import org.gradle.tooling.internal.consumer.DefaultGradleConnector
 import pprint.Util.literalize
 
+import java.io.File
 import java.util.Properties
 import java.util.concurrent.TimeUnit
 import scala.util.Using
@@ -39,6 +40,12 @@ object MillGradleBuildGenMain {
     val gradleWorkspace = os.Path.expandUser(projectDir, os.pwd)
     val millWorkspace = os.pwd
 
+    val gradleWrapperProperties = {
+      val file = gradleWorkspace / "gradle/wrapper/gradle-wrapper.properties"
+      val properties = new Properties()
+      if (os.isFile(file)) Using.resource(os.read.inputStream(file))(properties.load)
+      properties
+    }
     val exportPluginJar = Using.resource(
       getClass.getResourceAsStream(exportpluginAssemblyResource)
     )(os.temp(_, suffix = ".jar"))
@@ -60,6 +67,23 @@ object MillGradleBuildGenMain {
         conn
       case conn => conn
     }
+    if (gradleWrapperProperties.getProperty("distributionUrl") == null) {
+      // Fallback to system Gradle installation instead of the version corresponding to the
+      // Tooling API dependency.
+      System.getenv("GRADLE_HOME") match {
+        case null =>
+          os.proc("gradle", "--no-daemon", "--version").call().out.lines().collectFirst {
+            case s"Gradle ${gradleVersion}" =>
+              println(s"using Gradle version $gradleVersion")
+              gradleConnector.useGradleVersion(gradleVersion)
+          }.getOrElse {
+            sys.error(s"Failed to determine Gradle version. Please set GRADLE_HOME and retry.")
+          }
+        case gradleHome =>
+          println(s"using Gradle home $gradleHome")
+          gradleConnector.useInstallation(new File(gradleHome))
+      }
+    }
     var packages =
       try Using.resource(gradleConnector.forProjectDirectory(gradleWorkspace.toIO).connect) {
           connection =>
@@ -76,13 +100,9 @@ object MillGradleBuildGenMain {
       if (noMeta.value) (None, packages)
       else buildGen.withBaseModule(packages, "MavenModule" -> "MavenTests")
         .fold((None, packages))((base, pkgs) => (Some(base), pkgs))
-    val millJvmOpts = {
-      val properties = Properties()
-      val file = gradleWorkspace / "gradle/wrapper/gradle-wrapper.properties"
-      if (os.isFile(file)) Using.resource(os.read.inputStream(file))(properties.load)
-      val prop = properties.getProperty("org.gradle.jvmargs")
-      if (prop == null) Nil else prop.trim.split("\\s+").toSeq
-    }
+    val millJvmOpts = Option(
+      gradleWrapperProperties.getProperty("org.gradle.jvmargs")
+    ).fold(Nil)(_.trim.split("\\s+").toSeq)
     buildGen.writeBuildFiles(
       baseDir = millWorkspace,
       packages = packages0,

